common/bytes.go

```diff
@@ -0,0 +1,152 @@
+package common
+
+import (
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Encoder is a global setting for all byte encoding
+// This is the default.  Please override in the main()/init()
+// of your program to change how byte slices are presented
+//
+// In addition to these implementation, you can also find
+// BTCEncoder and FlickrEncoder that use base58 variants in
+// github.com/tendermint/go-wire/data/base58
+var (
+	encoder       ByteEncoder = hexEncoder{}
+	HexEncoder                = hexEncoder{}
+	B64Encoder                = base64Encoder{base64.URLEncoding}
+	RawB64Encoder             = base64Encoder{base64.RawURLEncoding}
+)
+
+func SetByteEncoder(e ByteEncoder) {
+	if encoder == nil {
+		encoder = e
+	} else {
+		panic("ByteEncoder already set.")
+	}
+}
+
+func GetByteEncoder() ByteEncoder {
+	return encoder
+}
+
+// For testing.
+func clearByteEncoder() {
+	encoder = nil
+}
+
+// Bytes is a special byte slice that allows us to control the
+// serialization format per app.
+//
+// Thus, basecoin could use hex, another app base64, and a third
+// app base58...
+type Bytes []byte
+
+// Marshal needed for protobuf compatibility
+func (b Bytes) Marshal() ([]byte, error) {
+	return b, nil
+}
+
+// Unmarshal needed for protobuf compatibility
+func (b *Bytes) Unmarshal(data []byte) error {
+	*b = data
+	return nil
+}
+
+func (b Bytes) MarshalJSON() ([]byte, error) {
+	return encoder.Marshal(b)
+}
+
+func (b *Bytes) UnmarshalJSON(data []byte) error {
```

Comment:
I definitely could be wrong but here are some thoughts:
I am not convinced that we should have a default codec and then use it in here magically on behalf of the user.
Why not let the user call say `HexCodec.UnmarshalJSON(data)`? What happens when we have two different programs that each import this package but might want to differently encode Byte?

For example if I have a struct with the address stored in hex but the digest, stored in base64
```go
type Data struct {
    Address cmn.Byte `json:"addr"` // <- to be stored in hex
    PassphraseDigest cmn.Byte `json:"pass_addr"` // <-- to be stored in Base64
}
```

then oops in order to successfully encode/decode that struct, I will have to implement a custom MarshalJSON, UnmarshalJSON that has to set the codec after each field has been traversed or invoke a separate codec for each field. Moreover this most likely would be done concurrently encoding and decoding other structs. It also would require the user to be an advanced Go user to successfully use the data types thus a higher mental load on the user. For example, this is what I think it will take for a user to successfully marshal the struct defined above that has fields that both are of type cmn.Bytes but they want the representations to be different on the wire:
```go
func (d *Data) MarshalJSON() ([]byte, error) {
   var err error
   repr := make(map[string]interface{})
   if repr["addr"], err = cmn.HexCodec.MarshalJSON(repr.Address); err != nil {
      return nil, err
   }
   if repr["pass_digest"], err = cmn.Base64Codec.MarshalJSON(repr.PassphraseDigest); err != nil {
      return nil, err
   }
   // Then for the other fields of the Data struct
   overallMap := make(map[string]interface{})
   overallBlob, err := json.Marshal(d)
   if err != nil {
      return nil, err
   }
   if err := json.Unmarshal(overallBlob, &overallMap); err != nil {
      return nil, err
   }
   // Now overwrite the keys of the custom base64 and hex keys encoded differently together
   for k, v := range repr {
       overallMap[k] = v
   }
   return json.Marshal(overallMap)
}
```
and notice how that requires advanced knowledge of the encoding/json package but also that the complexity increases with every single custom field added plus it'll do double work in the step of encoding the non-cmn.Bytes fields

### Proposed change
I defined separate types with different codecs, I now only need to define the struct as:
```go
type Data struct {
   Address cmn.HexBytes `json:"addr"`
   PassphraseDigest cmn.Base64Bytes `json:"pass_digest"`
}
```

Instead of implementing *Coder, perhaps we couldimplement different *Byte types, each with its respective *Marshal* and String methods. This way the user doesn't have to first set a global codec everytime, plus they now have better control in structs.
 
```go
type JSONBytes []byte
var _ json.Codec = (*JSONBytes)(nil)
var _ fmt.Stringer = (*JSONBytes)(nil)
type HexBytes baseBytes
var _ json.Codec = (*HexBytes)(nil)
var _ fmt.Stringer = (*HexBytes)(nil)
type Base64Bytes baseBytes
var _ json.Codec = (*Base64Bytes)(nil)
var _ fmt.Stringer = (*Base64Bytes)(nil)

// Implement the respective *arshal* and String methods for each type
```